lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -154,21 +144,45 @@ impl InvertedIndex {
             // Empty request -> no matches
             return Box::new(vec![].into_iter());
         }
-        intersect_postings_iterator(postings)
+
+        match self {
+            InvertedIndex::Mutable(_index) => {
+                // in case of mutable index, deleted documents are removed from the postings
+                intersect_postings_iterator(postings, |_| true)
+            }
+            InvertedIndex::Immutable(index) => {
+                // in case of immutable index, deleted documents are still in the postings
+                let filter = move |idx| {
+                    matches!(
+                        index.point_documents_tokens.get(idx as usize),
+                        Some(Some(_))
+                    )
+                };
+                intersect_postings_iterator(postings, filter)
+            }
+        }
     }
 
     pub fn estimate_cardinality(
         &self,
         query: &ParsedQuery,
         condition: &FieldCondition,
     ) -> CardinalityEstimation {
+        let points_count = match self {
+            InvertedIndex::Mutable(index) => index.points_count,
+            InvertedIndex::Immutable(index) => index.points_count,
+        };
+        let index_postings = match self {
+            InvertedIndex::Mutable(index) => &index.postings,
+            InvertedIndex::Immutable(index) => &index.postings,
+        };
```

Comment:
We can do a single match here that returns a tuple, like:

```rust
let (points_count, index_postings) = match self {
    InvertedIndex::Mutable(index) => (index.points_count, &index.postings),
    InvertedIndex::Immutable(index) => (index.points_count, &index.postings),
};
```

The same goes for some lines below.

If you like the current a lot better though, feel free to keep it as-is.